### ISSUE TITLE
Diamond Drill Recipe Broken

### ISSUE THREAD

**Issue by ShinoGFX** [GTNewHorizons/GT-New-Horizons-Modpack#6562]
#### Which modpack version are you using?

2.0.9.0

#### What did you try to do, and what did you expect to happen?

Attempted to craft the diamond chainsaw from the thaumcraft mods, I expected to be able to craft it. Right research, correct items, everything. But couldn’t craft it at all. Even in a creative world with everything unlocked quest achievement and research wise

#
#### What do you suggest instead/what changes do you propose?

Look into this and make the recipe actually craftable/Fix the crafting recipe. Thanks

**Comment by GTNH-Afx237v7** [comment]
Can you attach the screenshots you showed in Discord?
The diamond chainsaw is from Electro Magic Tools. Recipe is changed via Minetweaker script. Requires an MV electric chainsaw, seemingly needs to be made out of titanium with thaumium chainsaw tip. Running this chainsaw through a name remover didn’t help either. I’m not too experienced with MT scripts, maybe the issue lies in the definition of required metadata of the electric chainsaw.

**Comment by ShinoGFX** [comment]
![A644806D-C342-4839-A3C6-64C7AE5467F1](https://user-images.githubusercontent.com/48395343/91722952-a8d2da80-eb60-11ea-8242-bccbbd68733d.png)
![Uploading BC9EAEB1-32BD-4707-80DC-960743431C71.png…]()

**Comment by ShinoGFX** [comment]
The wand is in there just for some reason isn’t showing. I’ve had that visual glitch for a while.

**Comment by Prometheus0000** [comment]
This is the script:
// --- Diamond Chain Saw
mods.thaumcraft.Research.removeResearch("Diamond Chainsaw");
mods.thaumcraft.Research.addResearch("DiamondChainsaw", "EMT", "lucrum 12, ignis 9, metallum 6, terra 3", 0, -2 as int, 2, <EMT:DiamondChainsaw:27>.withTag({shearsMode: 1 as byte}));
mods.thaumcraft.Research.addPrereq("DiamondChainsaw", "ElectricMagicTools", false);
mods.thaumcraft.Research.setRound("DiamondChainsaw", true);
game.setLocalization("tc.research_name.DiamondChainsaw", "Diamond Chainsaw");
game.setLocalization("tc.research_text.DiamondChainsaw", "Harder, better, faster, stronger");
mods.thaumcraft.Research.addPage("DiamondChainsaw", "tc.research_page.DiamondChainsaw");
game.setLocalization("tc.research_page.DiamondChainsaw", "You are unsure why the thought of upgrading a Chainsaw with diamonds similarly to what you have done with a Drill has never occurred to you. This upgrade makes the Chainsaw capable of holding more energy and breaking blocks faster.");
mods.thaumcraft.Arcane.addShaped("DiamondChainsaw", <EMT:DiamondChainsaw:27>.withTag({shearsMode: 1 as byte}), "aer 35, terra 35, ignis 35, aqua 35, ordo 35, perditio 35", [
[<ore:screwDiamond>, <ore:plateDiamond>, <ore:screwDiamond>],
[<ore:plateDiamond>, <gregtech:gt.metatool.01:112>.withTag({ench: [{lvl: 2 as short, id: 35 as short}, {lvl: 2 as short, id: 21 as short}], "GT.ToolStats": {PrimaryMaterial: "Thaumium", SpecialData: -1 as long, MaxDamage: 51200 as long, Tier: 2 as long, MaxCharge: 400000 as long, Voltage: 128 as long, Electric: 1 as byte, SecondaryMaterial: "Titanium"}, "GT.CraftingComponents": {"Ingredient.5": {id: 7042 as short, Count: 1 as byte, Damage: 20028 as short}, "Ingredient.6": {id: 7041 as short, Count: 1 as byte, Damage: 17028 as short}, "Ingredient.7": {id: 7041 as short, Count: 1 as byte, Damage: 32528 as short}, "Ingredient.8": {id: 7041 as short, Count: 1 as byte, Damage: 17028 as short}, "Ingredient.0": {id: 7041 as short, Count: 1 as byte, Damage: 27028 as short}, "Ingredient.1": {id: 7042 as short, Count: 1 as byte, Damage: 9330 as short}, "Ingredient.3": {id: 7042 as short, Count: 1 as byte, Damage: 20028 as short}, "Ingredient.4": {id: 7041 as short, Count: 1 as byte, Damage: 32601 as short}}, "GT.ItemCharge": 400000 as long}), <ore:plateDiamond>],
[<ore:craftingToolHardHammer>, <ore:plateDiamond>, <ore:craftingToolWrench>]]);
mods.thaumcraft.Research.addArcanePage("DiamondChainsaw", <EMT:DiamondChainsaw:27>.withTag({shearsMode: 1 as byte}));

The important part is probably: 
<gregtech:gt.metatool.01:112>.withTag({ench: [{lvl: 2 as short, id: 35 as short}, {lvl: 2 as short, id: 21 as short}], "GT.ToolStats": {PrimaryMaterial: "Thaumium", SpecialData: -1 as long, MaxDamage: 51200 as long, Tier: 2 as long, MaxCharge: 400000 as long, Voltage: 128 as long, Electric: 1 as byte, SecondaryMaterial: "Titanium"}, "GT.CraftingComponents": {"Ingredient.5": {id: 7042 as short, Count: 1 as byte, Damage: 20028 as short}, "Ingredient.6": {id: 7041 as short, Count: 1 as byte, Damage: 17028 as short}, "Ingredient.7": {id: 7041 as short, Count: 1 as byte, Damage: 32528 as short}, "Ingredient.8": {id: 7041 as short, Count: 1 as byte, Damage: 17028 as short}, "Ingredient.0": {id: 7041 as short, Count: 1 as byte, Damage: 27028 as short}, "Ingredient.1": {id: 7042 as short, Count: 1 as byte, Damage: 9330 as short}, "Ingredient.3": {id: 7042 as short, Count: 1 as byte, Damage: 20028 as short}, "Ingredient.4": {id: 7041 as short, Count: 1 as byte, Damage: 32601 as short}}, "GT.ItemCharge": 400000 as long})

Which should be the drill.

**Comment by ShinoGFX** [comment]
Ive tried it even with all of that and it still does not work.

**Comment by GTNH-Afx237v7** [comment]
Crafting the electric chainsaw and then doing /mt Hand gives the following info:
<gregtech:gt.metatool.01:112>.withTag({ench: [{lvl: 2 as short, id: 35 as short}, {lvl: 2 as short, id: 21 as short}], "GT.ToolStats": {PrimaryMaterial: "Thaumium", SpecialData: -1 as long, MaxDamage: 51200 as long, Tier: 2 as long, MaxCharge: 400000 as long, Voltage: 128 as long, Electric: 1 as byte, SecondaryMaterial: "Titanium"}, "GT.CraftingComponents": {"Ingredient.5": {id: 6853 as short, Count: 1 as byte, registryName: "gregtech:gt.metaitem.02", Damage: 20028 as short}, "Ingredient.6": {id: 6852 as short, Count: 1 as byte, registryName: "gregtech:gt.metaitem.01", Damage: 17028 as short}, "Ingredient.7": {id: 6852 as short, Count: 1 as byte, registryName: "gregtech:gt.metaitem.01", Damage: 32528 as short}, "Ingredient.8": {id: 6852 as short, Count: 1 as byte, registryName: "gregtech:gt.metaitem.01", Damage: 17028 as short}, "Ingredient.0": {id: 6852 as short, Count: 1 as byte, registryName: "gregtech:gt.metaitem.01", Damage: 27028 as short}, "Ingredient.1": {id: 6853 as short, Count: 1 as byte, registryName: "gregtech:gt.metaitem.02", Damage: 9330 as short}, "Ingredient.3": {id: 6853 as short, Count: 1 as byte, registryName: "gregtech:gt.metaitem.02", Damage: 20028 as short}, "Ingredient.4": {id: 6852 as short, Count: 1 as byte, tag: {"GT.CraftingComponents": {"Ingredient.5": {id: 1480 as short, Count: 1 as byte, registryName: "gregtech:gt.blockmachines", Damage: 1341 as short}, "Ingredient.6": {id: 6852 as short, Count: 1 as byte, registryName: "gregtech:gt.metaitem.01", Damage: 23019 as short}, "Ingredient.7": {id: 1480 as short, Count: 1 as byte, registryName: "gregtech:gt.blockmachines", Damage: 1341 as short}, "Ingredient.8": {id: 1480 as short, Count: 1 as byte, registryName: "gregtech:gt.blockmachines", Damage: 1366 as short}, "Ingredient.0": {id: 1480 as short, Count: 1 as byte, registryName: "gregtech:gt.blockmachines", Damage: 1366 as short}, "Ingredient.1": {id: 1480 as short, Count: 1 as byte, registryName: "gregtech:gt.blockmachines", Damage: 1341 as short}, "Ingredient.2": {id: 6852 as short, Count: 1 as byte, registryName: "gregtech:gt.metaitem.01", Damage: 23019 as short}, "Ingredient.3": {id: 1480 as short, Count: 1 as byte, registryName: "gregtech:gt.blockmachines", Damage: 1341 as short}, "Ingredient.4": {id: 6852 as short, Count: 1 as byte, registryName: "gregtech:gt.metaitem.01", Damage: 23355 as short}}}, registryName: "gregtech:gt.metaitem.01", Damage: 32601 as short}}, "GT.ItemCharge": 400000 as long})

[...] {"Ingredient.5": {id: 6853 [...]
is where it begins being off from the Minetweaker script, as the script requires
[...] {"Ingredient.5": {id: 7042 [...]

Similar for all the remaining {"Ingredient.x":{id: xxxx)

I don't know how ID allocation works, or if there is a way around in this case?
Running the chainsaw through a Name remover leaves only

<gregtech:gt.metatool.01:112>.withTag({ench: [{lvl: 2 as short, id: 35 as short}, {lvl: 2 as short, id: 21 as short}], "GT.ToolStats": {PrimaryMaterial: "Thaumium", SpecialData: -1 as long, MaxDamage: 51200 as long, Tier: 2 as long, MaxCharge: 400000 as long, Voltage: 128 as long, Electric: 1 as byte, SecondaryMaterial: "Titanium"}, "GT.ItemCharge": 400000 as long})

but changing the Minetweaker script to this would require any craft to go through a Name remover first...

**Comment by Prometheus0000** [comment]
BTW you should be able to use NEI to dump the inventory data to a file instead of having to type all that out. Though I'm not certain it would list all the NBT data or not.

**Comment by GTNH-Afx237v7** [comment]
Copied from the latest.log file. 
For me to understand, the IDs of the ingredients (e.g. small titanium gear) don’t match. I spawned in the item with ID 7042 as indicated in the script and it gave me a piece of mini charcoal.
If someone else were to spawn in ID 7042, I assume they would receive a different item?

**Comment by Glease** [comment]
The crafting components tags is totally irrelevant here. I think removing that would fix the problem, but I don't have to time to test it out. Can anyone help test it out?

**Comment by GTNH-Afx237v7** [comment]
I can test tomorrow. What exactly should be removed? Everything after
<gregtech:gt.metatool.01:112>
?

**Comment by Glease** [comment]
Replace the ingredient with 
```zs
<gregtech:gt.metatool.01:112>.withTag(
  {
    ench: [
      {
        lvl: 2 as short,
        id: 35 as short
      },
      {
        lvl: 2 as short,
        id: 21 as short
      }
    ],
    "GT.ToolStats": {
      PrimaryMaterial: "Thaumium",
      SpecialData: -1 as long,
      MaxDamage: 51200 as long,
      Tier: 2 as long,
      MaxCharge: 400000 as long,
      Voltage: 128 as long,
      Electric: 1 as byte,
      SecondaryMaterial: "Titanium"
    },
    "GT.ItemCharge": 400000 as long
  }
)
```
Or in short, remove `GT.CraftingComponents` subtag.

**Comment by GTNH-Afx237v7** [comment]
Ok thx, will test tomorrow and let you know.

**Comment by GTNH-Afx237v7** [comment]
Pro: with the change suggested by Glease, you can now craft the diamond chainsaw
Con: the electric chainsaw is not consumed, instead remains in the arcane crafting grid

**Comment by Glease** [comment]
I guess it's because it's considered a crafting tool (like hammer/wrench) etc. I have no idea if we can get around this via MT. If not, maybe we can just replace the complete chainsaw with a titanium saw head and add in a MV battery unit.

**Comment by GTNH-Afx237v7** [comment]
Don’t think it’ll work but I’ll give the no.return() function a try tomorrow. If that doesn’t solve it I’ll make a recipe suggestion. Will probably need adjustments then for the other tools of this mod as well.

**Comment by Glease** [comment]
Theoretically it should work. Technically the relationship between the new chainsaw (with power consumed and durability reduced) and the original chainsaw is identical to the relationship betwen an empty bucket and a water bucket, so `.noReturn()` should work.

**Comment by GTNH-Afx237v7** [comment]
Theoretically.... most likely my fault, but when adding the .noReturn() the recipe changes completely....
![grafik](https://user-images.githubusercontent.com/64365566/92082758-43cbee80-edc5-11ea-94e4-caf618932745.png)

Here is how I changed it:
`<gregtech:gt.metatool.01:112>.withTag({ench: [{lvl: 2 as short, id: 35 as short}, {lvl: 2 as short, id: 21 as short}], "GT.ToolStats": {PrimaryMaterial: "Thaumium", SpecialData: -1 as long, MaxDamage: 51200 as long, Tier: 2 as long, MaxCharge: 400000 as long, Voltage: 128 as long, Electric: 1 as byte, SecondaryMaterial: "Titanium"}, "GT.ItemCharge": 400000 as long}.noReturn())`

I first tried
`<gregtech:gt.metatool.01:112>.withTag({ench: [{lvl: 2 as short, id: 35 as short}, {lvl: 2 as short, id: 21 as short}], "GT.ToolStats": {PrimaryMaterial: "Thaumium", SpecialData: -1 as long, MaxDamage: 51200 as long, Tier: 2 as long, MaxCharge: 400000 as long, Voltage: 128 as long, Electric: 1 as byte, SecondaryMaterial: "Titanium"}, "GT.ItemCharge": 400000 as long}).noReturn()`

but putting the **)** before .noReturn() caused the recipe to not being shown in NEI at all...

**Comment by GTNH-Afx237v7** [comment]
I searched around the internet a bit, came up with two attempts:

1/ .transformDamage().noReturn() resulting in
ERROR: Error executing Emt.zs: Invalid shaped ore recipe: abc, def, ghi, a, screwDiamond, b, plateDiamond, c, screwDiamond, d, plateDiamond, e, null, f, plateDiamond, g, craftingToolHardHammer, h, plateDiamond, i, craftingToolWrench, 1xitem.EMT.chainsaw.diamond@27

2/ .transformConsume() resulting in
ERROR: Emt.zs:224 > a method available but none matches the parameters ()
This is usually an error in your script, not in the mod
transformConsume(ZenTypeNative: minetweaker.item.IIngredient, int)

**Comment by GTNH-Afx237v7** [comment]
@Glease are you able to check the Minetweaker code and find out what possibilities there are for the .transform() function? Unfortunately, my programming experience equals pretty much zero... maybe there is something in there that allows the full consumption of a GT metatool...

**Comment by Glease** [comment]
The problem seems to be the MT code itself. noReturn seems to doing something really weird. It returns an empty value instead of the altered ingredient. I googled a bit and so far I see no similiar complains. On the other hand transformConsume requires an integer as parameter, so you need to write `transformConsume(1)` instead of just transformConsume(). Here the 1 inside the parenthesis means one item will be consumed, which is appropriate in our case.

I'll try to find some time and see if I can find the cause of the noReturn problem, but meanwhile please try the `transformConsume(1)` path and see if it works.

**Comment by GTNH-Afx237v7** [comment]
Does that help/give info in any way?

https://github.com/stanhebben/MineTweaker3/issues/336

**Comment by Glease** [comment]
Nope. They are saying the balky craft handler is not supporting the given ingredient, but we here have a balky ingredient that refuse to add the transform correctly. BTW can you please try `<gregtech:meta...:..>.withTag({...}).transformConsume(1)`?

**Comment by GTNH-Afx237v7** [comment]
`<gregtech:gt.metatool.01:112>.withTag({ench: [{lvl: 2 as short, id: 35 as short}, {lvl: 2 as short, id: 21 as short}], "GT.ToolStats": {PrimaryMaterial: "Thaumium", SpecialData: -1 as long, MaxDamage: 51200 as long, Tier: 2 as long, MaxCharge: 400000 as long, Voltage: 128 as long, Electric: 1 as byte, SecondaryMaterial: "Titanium"}, "GT.ItemCharge": 400000 as long}).transformConsume(1),` 
unfortunately doesn't work; neither does `.withAmount(1).transformConsume(1),`

Both result in the following error message:

ERROR: Error executing Emt.zs: Invalid shaped ore recipe: abc, def, ghi, a, screwDiamond, b, plateDiamond, c, screwDiamond, d, plateDiamond, e, null, f, plateDiamond, g, craftingToolHardHammer, h, plateDiamond, i, craftingToolWrench, 1xitem.EMT.chainsaw.diamond@27

i.e. it is now not registering the electric chainsaw at all.

**Comment by leagris** [comment]
Did you try using Function Recipes as demonstrated in [minetweaker3.powerofbytes.com: Advanced Recipes / Function Recipes](http://minetweaker3.powerofbytes.com/wiki/Tutorial:Advanced_Recipes)?

```zenscript
mods.thaumcraft.Arcane.addShaped(
  "DiamondChainsaw",
  <EMT:DiamondChainsaw:27>.withTag({shearsMode: 1 as byte}),
  "aer 35, terra 35, ignis 35, aqua 35, ordo 35, perditio 35",
  [
    [<ore:screwDiamond>, <ore:plateDiamond>, <ore:screwDiamond>],
    [<ore:plateDiamond>, <gregtech:gt.metatool.01:112>.withTag({ench: [{lvl: 2 as short, id: 35 as short}, {lvl: 2 as short, id: 21 as short}], "GT.ToolStats": {PrimaryMaterial: "Thaumium", SpecialData: -1 as long, MaxDamage: 51200 as long, Tier: 2 as long, MaxCharge: 400000 as long, Voltage: 128 as long, Electric: 1 as byte, SecondaryMaterial: "Titanium"}, "GT.CraftingComponents": {"Ingredient.5": {id: 7042 as short, Count: 1 as byte, Damage: 20028 as short}, "Ingredient.6": {id: 7041 as short, Count: 1 as byte, Damage: 17028 as short}, "Ingredient.7": {id: 7041 as short, Count: 1 as byte, Damage: 32528 as short}, "Ingredient.8": {id: 7041 as short, Count: 1 as byte, Damage: 17028 as short}, "Ingredient.0": {id: 7041 as short, Count: 1 as byte, Damage: 27028 as short}, "Ingredient.1": {id: 7042 as short, Count: 1 as byte, Damage: 9330 as short}, "Ingredient.3": {id: 7042 as short, Count: 1 as byte, Damage: 20028 as short}, "Ingredient.4": {id: 7041 as short, Count: 1 as byte, Damage: 32601 as short}}, "GT.ItemCharge": 400000 as long}), <ore:plateDiamond>],
    [<ore:craftingToolHardHammer>, <ore:plateDiamond>, <ore:craftingToolWrench>]
  ],
  function(output, inputs, crafting) {
    return output;
  }
);
```

**Comment by GTNH-Afx237v7** [comment]
I didn’t as I don’t have the knowledge... if you make a suggestion I can test again and let you know.

**Comment by leagris** [comment]
> if you make a suggestion I can test again and let you know.

I thought [this](#issuecomment-687598048) was a syntactically correct Function Recipe suggestion, with the recipe function returning only the output.

I am not sure if the Thaumcraft's Arcane crafting table Modtweaker's interface implements all the advanced features of the MineTweaker's workbench recipes though!

**Comment by GTNH-Afx237v7** [comment]
I will try your suggestion and let you know.

**Comment by GTNH-Afx237v7** [comment]
Again, no luck...
ERROR: Emt.zs:222 > a method available but none matches the parameters (string, ZenTypeNative: minetweaker.item.IItemStack, string, any[], function(any,any,any)any)
This is usually an error in your script, not in the mod
addShaped(string, ZenTypeNative: minetweaker.item.IItemStack, string, ZenTypeNative: minetweaker.item.IIngredient[][])

**Comment by leagris** [comment]
it show that ModTweaker's Thaumcraft interface does not implement the whole set of features of the vanilla Crafting Table's MineTweaker interface, or that Thaumcraft's Arcane Crafting Table lacks the needed methods and logic to handle CraftTweaker's Function Recipes and Ingredient return modifiers.

**Comment by GTNH-Afx237v7** [comment]
Seems like it, all in all I attempted it nine times with no eventually acceptable result, i.e. the consumption of the GT electric chainsaw...

**Comment by GTNH-Afx237v7** [comment]
Here my suggestion for a changed recipe:
![grafik](https://user-images.githubusercontent.com/64365566/92392230-3d6ba880-f11e-11ea-86de-1d04559ac73f.png)

Titanium Screw/Thaumium Chainsaw Tip/Titanium Screw
Diamond Plate/HV-Tier circuit (Oredict enabled)/Diamond Plate
GT Tool Hammer/Empty MV Lithium Battery/GT Tool Wrench

Aspects remain the same

**Comment by Prometheus0000** [comment]
This is fine with me, but you should also check the other tool ones afterward, no point fixing one and leaving others broken.

**Comment by GTNH-Afx237v7** [comment]
I did test upgrading the diamond chainsaw into a thaumium chainsaw, worked fine. Beyond the above mentioned items, the other recipes are infusion recipes.
Looking at the scripts, some may need adjustment, however I currently lack the time to test all. Will come back later.

**Comment by Prometheus0000** [comment]
Related: #2694. I fixed the circuits oredict earlier.

**Comment by Prometheus0000** [comment]
So this is done with the PR? @GTNH-Afx237v7

**Comment by GTNH-Afx237v7** [comment]
Just stumbled across this. Yes the PR fixes this, I tested it with 2.0.9.0 (end of September 2020) before making the PR.
